### PR TITLE
Add option to skip masking of edge tubes in all banks in IntegratePeaksMD

### LINKED ISSUE #31008: IntegratePeaksMD edge check not working for WISH data
### Expected behavior
IntegratePeaksMD would identify peaks on the detector edge so they can be excluded before refinement.

### Actual behavior
Existing solution currently identifies many peaks as being on the edge that aren't - in addition the existing code has some Correli masking. If instrument specific masking is needed then it should not live in the algorithm but perhaps in the instrument definition?

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Description of work.**

Previously IntegratePeaksMD would (for instruments other than Correlli) perform two masking operations to signify which pixels were on the edge so peaks can be excluded if not completely on the detector.
(1) Mask the pixels at the top and bottom of each tube
(2) Mask the first and last tubes in each bank

For example on WISH this results in
image
However, this masks tubes between adjacent banks (see above) which we do not want to consider edges on WISH. For Corelli there is instrument specific code on
https://github.com/mantidproject/mantid/blob/f643296c46e4e95bcc4e747612842c7b2e37c0f4/Framework/MDAlgorithms/src/IntegratePeaksMD2.cpp#L1154
that applies a custom masking to prevent this problem - however a more general solution is to apply a mask on the peaks workspace prior to integration.

To this end, I have added an option to stop masking the edge tubes - i.e. make masking operation (2) optional . By default they are still masked so nothing breaks for Correlli. WISH will then apply a custom masking of tubes prior to integration (the pixel masking is left in at present as this was not Corelli specific and works for WISH and typically tube ends are less reliable anyway).

Long term I think the Corelli specific code should be removed in favour of all instrument applying custom masking before integration (i.e. in the workflow script not the algorithm) - but WISH need this working in this release and I didn't want to break anything for Corelli.

I have also renamed the function that calculated the distance  of a peak Q and fixed bug such that now we look for the minimum distance - i.e. the detector trajectory closest to peak Q - this is important for the correction of the intensity if on edge (although this only works for spherical peaks currently so have updated the docs).

It is not possible at present to check masking and edge identification/correction in the unit tests IntegratePEaksMD2Test.h as MaskBTP cannot be called as it is a python algorithm - this will be the subject of a separate issue.
For now I have tested the edge masking in a WISH SX reduction system test which has the advantage of adding coverage of  algorithms that are important to their workflow (CreatePeaksWorkspace, IntegratePeaksMD and MaskBTP)

**To test:**
(1) Follow this script and check the correct peaks are integrated (data on my babylon - special cropped MD workspace that takes up less memory)

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

LoadMD(Filename='//olympic/Babylon5/Public/RWaite/WISH37185MD_QLab_crop.nxs', OutputWorkspace='MD')

# Mask just pixels at end of each tube - this should not integrate peak index 3
LoadNexus(Filename='//olympic/Babylon5/Public/RWaite/EdgeCheckPeaksWISH.nxs', OutputWorkspace='peaks')
IntegratePeaksMD(InputWorkspace='MD', PeakRadius='0.02', PeaksWorkspace='peaks', 
    OutputWorkspace='peaks_int', IntegrateIfOnEdge=False, MaskEdgeTubes=False) 
# Mask just pixels at end of each tube - this should not integrate peak indices 0,2,3
IntegratePeaksMD(InputWorkspace='MD', PeakRadius='0.02', PeaksWorkspace='peaks', 
    OutputWorkspace='peaks_int', IntegrateIfOnEdge=False, MaskEdgeTubes=True) 

# Need to reload peaks workspace for next test (so no masking has been applied)
LoadNexus(Filename='//olympic/Babylon5/Public/RWaite/EdgeCheckPeaksWISH.nxs', OutputWorkspace='peaks')
# apply custom masking of tubes on WISH
MaskBTP(Workspace='peaks', Bank='5-6', Tube='152')
MaskBTP(Workspace='peaks', Bank='1,10', Tube='1')
# Should not integrate peak indices 0, 3
IntegratePeaksMD(InputWorkspace='MD', PeakRadius='0.02', PeaksWorkspace='peaks', 
    OutputWorkspace='peaks_int', IntegrateIfOnEdge=False, MaskEdgeTubes=False) 
```
 Here is a image showing the peak positions for reference
image

Fixes #31008 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/MDAlgorithms/inc/MantidMDAlgorithms/IntegratePeaksMD2.h [hunk 1]
```diff
@@ -79,7 +79,7 @@ class DLLExport IntegratePeaksMD2 : public API::Algorithm {
 
   /// Calculate if this Q is on a detector
   void calculateE1(const Geometry::DetectorInfo &detectorInfo);
-  double detectorQ(Mantid::Kernel::V3D QLabFrame, double r);
+  double calculateDistanceToEdge(Mantid::Kernel::V3D QLabFrame, double r);
```
Comment (DannyHindson) [inline Framework/MDAlgorithms/inc/MantidMDAlgorithms/IntegratePeaksMD2.h:82]: This predates your changes but the QLabFrame parameter should be const reference I think. The function itself should also be labelled const (at the end)

File: Framework/MDAlgorithms/src/IntegratePeaksMD2.cpp [hunk 2]
```diff
@@ -178,6 +179,13 @@ void IntegratePeaksMD2::init() {
                   "Number of iterations in covariance estimation (ignored if all "
                   "peak radii are specified). 2-3 should be sufficient.");
 
+  declareProperty(
+      "MaskEdgeTubes", true,
+      "Mask tubes on the edge of all banks in the PeaksWorkspace instrument (note the edge pixels at top/bottom of all "
+      "tubes will always be masked even if this property is False). Note the algorithm will treat "
+      "any masked pixels as edges (including pixles already masked prior to the execution of this algorithm) - this "
```
Comment (DannyHindson) [inline Framework/MDAlgorithms/src/IntegratePeaksMD2.cpp:186]: typo in pixels

File: Framework/MDAlgorithms/src/IntegratePeaksMD2.cpp [hunk 3]
```diff
@@ -1213,17 +1226,15 @@ void IntegratePeaksMD2::calculateE1(const Geometry::DetectorInfo &detectorInfo)
...
+double IntegratePeaksMD2::calculateDistanceToEdge(Mantid::Kernel::V3D QLabFrame, double r) {
+  double edgeDist = r;
   for (auto &E1 : E1Vec) {
     V3D distv = QLabFrame - E1 * (QLabFrame.scalar_prod(E1)); // distance to the
                                                               // trajectory as a
                                                               // vector
-    if (distv.norm() < r) {
-      edge = distv.norm();
-    }
+    edgeDist = std::min(edgeDist, distv.norm());              // want smallest dist to
```
Comment (DannyHindson) [inline Framework/MDAlgorithms/src/IntegratePeaksMD2.cpp:1235]: should the comment have an extra word on the end ("want smallest dist to...")?

File: Framework/MDAlgorithms/test/IntegratePeaksMD2Test.h [hunk 4]
```diff
@@ -12,6 +12,7 @@
 #include "MantidAPI/Run.h"
 #include "MantidDataObjects/LeanElasticPeaksWorkspace.h"
 #include "MantidDataObjects/MDEventFactory.h"
+#include "MantidDataObjects/NoShape.h"
```
Comment (DannyHindson) [inline Framework/MDAlgorithms/test/IntegratePeaksMD2Test.h:15]: What's this extra header for? Not sure this is required

File: Framework/MDAlgorithms/src/IntegratePeaksMD2.cpp [hunk 5]
```diff
@@ -477,11 +490,11 @@ template <typename MDE, size_t nd> void IntegratePeaksMD2::integrate(typename MD
 
     // Do not integrate if sphere is off edge of detector
 
-    double edge = detectorQ(p.getQLabFrame(), std::max(BackgroundOuterRadius[0], PeakRadius[0]));
-    if (edge < std::max(BackgroundOuterRadius[0], PeakRadius[0])) {
+    double edgeDist = calculateDistanceToEdge(p.getQLabFrame(), std::max(BackgroundOuterRadius[0], PeakRadius[0]));
```
Comment (DannyHindson) [inline Framework/MDAlgorithms/src/IntegratePeaksMD2.cpp:493]: I seem to have got v enthusiastic about my const qualifiers this evening! I think edgeDist should be declared const double. Again - predates your changes

**Review by DannyHindson — CHANGES_REQUESTED** [review]
I've run the test script and I see it excluding the peaks in the different parts of the test as required.   Code changes look good. Nice clarifications around some of the function\variable names. Have highlighted a few trivial bits and pieces so hopefully quick to fix on Monday

File: Framework/MDAlgorithms/test/IntegratePeaksMD2Test.h (hunk 4 above)
Comment (RichardWaiteSTFC) [reply Framework/MDAlgorithms/test/IntegratePeaksMD2Test.h:15]: Good spot, that was used for a test I tried to write before discovering MaskBTP wouldn't get called - I have removed it.

File: Framework/MDAlgorithms/inc/MantidMDAlgorithms/IntegratePeaksMD2.h (hunk 1 above)
Comment (RichardWaiteSTFC) [reply Framework/MDAlgorithms/inc/MantidMDAlgorithms/IntegratePeaksMD2.h:82]: Thanks, yes you are correct and I have also removed the input `r` as I can just initialise with `edgeDist` with `DBL_MAX`

File: Framework/MDAlgorithms/inc/MantidMDAlgorithms/IntegratePeaksMD2.h [hunk 6]
```diff
@@ -79,7 +79,7 @@ class DLLExport IntegratePeaksMD2 : public API::Algorithm {
 
   /// Calculate if this Q is on a detector
   void calculateE1(const Geometry::DetectorInfo &detectorInfo);
-  double detectorQ(Mantid::Kernel::V3D QLabFrame, double r);
+  double calculateDistanceToEdge(const Mantid::Kernel::V3D QLabFrame);
```
Comment (DannyHindson) [inline Framework/MDAlgorithms/inc/MantidMDAlgorithms/IntegratePeaksMD2.h:82]: Please can you change this to be passed by reference also?

**Review by DannyHindson — CHANGES_REQUESTED** [review]

**Review by DannyHindson — APPROVED** [review]
Looks good thanks

**Pull request merged**